File: app/src/main/java/top/itning/yunshuclassschedule/ui/adapter/TodayRecyclerViewAdapter.java
package top.itning.yunshuclassschedule.ui.adapter;

import android.content.Context;
import android.graphics.Point;
import android.util.Log;
import android.view.Display;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.FrameLayout;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Random;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import top.itning.yunshuclassschedule.R;
import top.itning.yunshuclassschedule.entity.ClassSchedule;
import top.itning.yunshuclassschedule.ui.view.RoundBackChange;
import top.itning.yunshuclassschedule.util.ClassScheduleUtils;
import top.itning.yunshuclassschedule.util.DateUtils;
import top.itning.yunshuclassschedule.util.ThemeChangeUtil;

/**
 * 今天课程列表适配器
 *
 * @author itning
 */
public class TodayRecyclerViewAdapter extends RecyclerView.Adapter {
    private static final String TAG = "TodayRecyclerAdapter";
    /**
     * 列表数据集合
     */
    private final List<ClassSchedule> scheduleList;
    /**
     * 颜色数组
     */
    private final int[] colorArray = new int[7];
    /**
     * {@link Context}
     */
    private final Context context;
    /**
     * 随机好的颜色集合
     */
    private final ArrayList<Integer> showColorList;
    private View viewProgress;

    public TodayRecyclerViewAdapter(@NonNull List<ClassSchedule> scheduleList, @NonNull Context context) {
        Log.d(TAG, "new Today Recycler View Adapter");
        this.context = context;
        this.scheduleList = scheduleList;
        //数组赋值
        colorArray[0] = ContextCompat.getColor(context, R.color.class_color_1);
        colorArray[1] = ContextCompat.getColor(context, R.color.class_color_2);
        colorArray[2] = ContextCompat.getColor(context, R.color.class_color_3);
        colorArray[3] = ContextCompat.getColor(context, R.color.class_color_4);
        colorArray[4] = ContextCompat.getColor(context, R.color.class_color_5);
        colorArray[5] = ContextCompat.getColor(context, R.color.class_color_6);
        colorArray[6] = ContextCompat.getColor(context, R.color.class_color_7);
        //随机颜色集合构建
        Random random = new Random();
        showColorList = new ArrayList<>(colorArray.length);
        do {
            int number = random.nextInt(colorArray.length);
            if (!showColorList.contains(number)) {
                showColorList.add(number);
            }
        } while (showColorList.size() != colorArray.length);
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        Log.d(TAG, "onCreateViewHolder");
        return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_class_rv, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        Log.d(TAG, "onBindViewHolder pos->" + position);
        ClassSchedule classSchedule = scheduleList.get(position);
        ViewHolder viewHolder = (ViewHolder) holder;
        viewHolder.tvName.setText(classSchedule.getName());
        viewHolder.tvLocation.setText(classSchedule.getLocation());
        viewHolder.tvTime.setText(DateUtils.getTimeList().get(classSchedule.getSection() - 1));
        viewHolder.round.setBackColor(colorArray[showColorList.get(position)]);
        //显示设置可见性
        viewHolder.flNo.setVisibility(View.VISIBLE);
        viewHolder.viewBottom.setVisibility(View.INVISIBLE);
        viewHolder.viewTop.setVisibility(View.INVISIBLE);
        viewHolder.viewLeft.setVisibility(View.INVISIBLE);
        viewHolder.viewProgress.setVisibility(View.INVISIBLE);
        ThemeChangeUtil.setProgressBackgroundResource(context, viewHolder.viewProgress);
        ThemeChangeUtil.setBackgroundResources(context, viewHolder.viewBottom, viewHolder.viewTop, viewHolder.viewLeft, viewHolder.viewCenter);
        if (position == 0 && ClassScheduleUtils.haveClassAfterTime(scheduleList)) {
            //是当前正在或要上的课程
            viewHolder.flNo.setVisibility(View.INVISIBLE);
            viewHolder.viewBottom.setVisibility(View.VISIBLE);
            viewHolder.viewTop.setVisibility(View.VISIBLE);
            viewHolder.viewLeft.setVisibility(View.VISIBLE);
            viewHolder.viewProgress.setVisibility(View.VISIBLE);
            Display display = ((WindowManager) Objects.requireNonNull(context.getSystemService(Context.WINDOW_SERVICE))).getDefaultDisplay();
            Point size = new Point();
            display.getSize(size);
            ViewGroup.LayoutParams layoutParams = viewHolder.viewProgress.getLayoutParams();
            layoutParams.width = DateUtils.getNowProgress(size.x, scheduleList);
            viewHolder.viewProgress.setLayoutParams(layoutParams);
            viewProgress = viewHolder.viewProgress;
        }
    }

    public View getViewProgress() {
        return viewProgress;
    }

    @Override
    public int getItemCount() {
        return scheduleList == null ? 0 : scheduleList.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.view_left)
        View viewLeft;
        @BindView(R.id.view_top)
        View viewTop;
        @BindView(R.id.view_bottom)
        View viewBottom;
        @BindView(R.id.view_center)
        View viewCenter;
        @BindView(R.id.view_progress)
        View viewProgress;
        @BindView(R.id.round)
        RoundBackChange round;
        @BindView(R.id.fl_no)
        FrameLayout flNo;
        @BindView(R.id.tv_name)
        TextView tvName;
        @BindView(R.id.tv_location)
        TextView tvLocation;
        @BindView(R.id.tv_time)
        TextView tvTime;

        ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }
}

File: app/src/main/java/top/itning/yunshuclassschedule/ui/adapter/TodayRecyclerViewAdapter.kt
package top.itning.yunshuclassschedule.ui.adapter

import android.content.Context
import android.graphics.Point
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.FrameLayout
import android.widget.TextView
import androidx.annotation.NonNull
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import top.itning.yunshuclassschedule.R
import top.itning.yunshuclassschedule.entity.ClassSchedule
import top.itning.yunshuclassschedule.ui.view.RoundBackChange
import top.itning.yunshuclassschedule.util.ClassScheduleUtils
import top.itning.yunshuclassschedule.util.DateUtils
import top.itning.yunshuclassschedule.util.ThemeChangeUtil
import java.util.*

/**
 * 今天课程列表适配器
 *
 * @author itning
 */
class TodayRecyclerViewAdapter(
        /**
         * 列表数据集合
         */
        @param:NonNull private val scheduleList: List<ClassSchedule>?,
        /**
         * [Context]
         */
        @param:NonNull private val context: Context) : RecyclerView.Adapter<TodayRecyclerViewAdapter.ViewHolder>() {
    /**
     * 颜色数组
     */
    private val colorArray = IntArray(7)
    /**
     * 随机好的颜色集合
     */
    private val showColorList: ArrayList<Int>
    var viewProgress: View? = null
        private set

    init {
        Log.d(TAG, "new Today Recycler View Adapter")
        //数组赋值
        colorArray[0] = ContextCompat.getColor(context, R.color.class_color_1)
        colorArray[1] = ContextCompat.getColor(context, R.color.class_color_2)
        colorArray[2] = ContextCompat.getColor(context, R.color.class_color_3)
        colorArray[3] = ContextCompat.getColor(context, R.color.class_color_4)
        colorArray[4] = ContextCompat.getColor(context, R.color.class_color_5)
        colorArray[5] = ContextCompat.getColor(context, R.color.class_color_6)
        colorArray[6] = ContextCompat.getColor(context, R.color.class_color_7)
        //随机颜色集合构建
        val random = Random()
        showColorList = ArrayList(colorArray.size)
        do {
            val number = random.nextInt(colorArray.size)
            if (!showColorList.contains(number)) {
                showColorList.add(number)
            }
        } while (showColorList.size != colorArray.size)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        Log.d(TAG, "onCreateViewHolder")
        return ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_class_rv, parent, false))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        Log.d(TAG, "onBindViewHolder pos->$position")
        val classSchedule = scheduleList!![position]
        holder.tvName.text = classSchedule.name
        holder.tvLocation.text = classSchedule.location
        holder.tvTime.text = DateUtils.timeList[classSchedule.section - 1]
        if (position >= showColorList.size) {
            val number = Random().nextInt(colorArray.size)
            holder.round.setBackColor(colorArray[number])
        } else {
            holder.round.setBackColor(colorArray[showColorList[position]])
        }
        //显示设置可见性
        holder.flNo.visibility = View.VISIBLE
        holder.viewBottom.visibility = View.INVISIBLE
        holder.viewTop.visibility = View.INVISIBLE
        holder.viewLeft.visibility = View.INVISIBLE
        holder.viewProgress.visibility = View.INVISIBLE
        ThemeChangeUtil.setProgressBackgroundResource(context, holder.viewProgress)
        ThemeChangeUtil.setBackgroundResources(context, holder.viewBottom, holder.viewTop, holder.viewLeft, holder.viewCenter)
        if (position == 0 && ClassScheduleUtils.haveClassAfterTime(scheduleList)) {
            //是当前正在或要上的课程
            holder.flNo.visibility = View.INVISIBLE
            holder.viewBottom.visibility = View.VISIBLE
            holder.viewTop.visibility = View.VISIBLE
            holder.viewLeft.visibility = View.VISIBLE
            holder.viewProgress.visibility = View.VISIBLE
            val display = (Objects.requireNonNull(context.getSystemService(Context.WINDOW_SERVICE)) as WindowManager).defaultDisplay
            val size = Point()
            display.getSize(size)
            val layoutParams = holder.viewProgress.layoutParams
            layoutParams.width = DateUtils.getNowProgress(size.x, scheduleList)
            holder.viewProgress.layoutParams = layoutParams
            viewProgress = holder.viewProgress
        }
    }

    override fun getItemCount(): Int {
        return scheduleList?.size ?: 0
    }

    class ViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val viewLeft: View = itemView.findViewById(R.id.view_left)

        val viewTop: View = itemView.findViewById(R.id.view_top)

        val viewBottom: View = itemView.findViewById(R.id.view_bottom)

        val viewCenter: View = itemView.findViewById(R.id.view_center)

        val viewProgress: View = itemView.findViewById(R.id.view_progress)

        val round: RoundBackChange = itemView.findViewById(R.id.round)

        val flNo: FrameLayout = itemView.findViewById(R.id.fl_no)

        val tvName: TextView = itemView.findViewById(R.id.tv_name)

        val tvLocation: TextView = itemView.findViewById(R.id.tv_location)

        val tvTime: TextView = itemView.findViewById(R.id.tv_time)
    }

    companion object {
        private const val TAG = "TodayRecyclerAdapter"
    }
}
